### PR TITLE
feat: implement CacheClient

### PR DESCRIPTION
**Pull request by anitarua**
Addresses #29 and #30

Creates the CacheClient with both control and data plane operations.

- Control plane: create, delete, list caches
- Data plane: scalar get and set

Additional changes:

- added test set up and teardown utility functions
- added [docstrings](https://nshipster.com/swift-documentation/) to the public clients (CacheClient and TopicClient) -- you can option+click on a function name to see what it will look like
- use API_KEY terminology instead of AUTH_TOKEN for the environment variables
- added the enum StringOrData to replace instances of overloaded functions

### CODE REVIEW

File: Tests/momentoTests/testUtils.swift [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+struct TestSetup {
+    var cacheName: String
+    var cacheClient: CacheClientProtocol
+    var topicClient: TopicClientProtocol
+}
+
+func setUpIntegrationTests() async -> TestSetup {
+    do {
+        let creds = try CredentialProvider.fromEnvironmentVariable(envVariableName: "TEST_AUTH_TOKEN")
```
Comment (malandis) [inline Tests/momentoTests/testUtils.swift:12]:
```suggestion
        let creds = try CredentialProvider.fromEnvironmentVariable(envVariableName: "TEST_API_KEY")
```

We are using the terminology API key now 👍

File: Tests/momentoTests/cacheTests.swift [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+            )
+        default:
+            XCTAssertTrue(
+                listResult is CacheListSuccess,
+                "Unexpected response: \(listResult)"
+            )
+        }
+        
+        // Create new cache
+        let newCacheName = "a-totally-different-cache"
```
Comment (malandis) [inline Tests/momentoTests/cacheTests.swift:82]: You may want to create a method to generate a random cache name. That way if you have multiple CI jobs running at once, then these names will be unique.

File: Tests/momentoTests/cacheTests.swift [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+            invalidKey is CacheGetError,
+            "Unexpected response: \(invalidKey)"
+        )
+        let invalidKeyErrorCode = (invalidKey as! CacheGetError).errorCode
+        XCTAssertEqual(
+            invalidKeyErrorCode, MomentoErrorCode.INVALID_ARGUMENT_ERROR,
+            "Unexpected error code: \(invalidKeyErrorCode)"
+        )
+        
+        let stringKey = await self.cacheClient.get(cacheName: self.integrationTestCacheName, key: "hello")
```
Comment (malandis) [inline Tests/momentoTests/cacheTests.swift:144]: Same recommendation applies here re random key names.

File: Tests/momentoTests/configTests.swift [hunk 4]
```diff
@@ -23,17 +40,16 @@ final class configTests: XCTestCase {
...
-//        let pubResp = await client.publish(
-//            cacheName: "test-cache",
-//            topicName: "test-topic",
-//            value: "test-message"
-//        )
-//        XCTAssertTrue(pubResp is TopicPublishError)
-//        XCTAssertEqual(MomentoErrorCode.TIMEOUT_ERROR, (pubResp as! TopicPublishError).errorCode)
-//    }
+    func testTimeoutForImpossibleDeadline() async throws {
+        let creds = try CredentialProvider.fromEnvironmentVariable(envVariableName: "TEST_AUTH_TOKEN")
```
Comment (malandis) [inline Tests/momentoTests/configTests.swift:44]: I recommend externalizing the env var name to a variable in one place

File: Sources/momento/CacheClient.swift [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+public class CacheClient: CacheClientProtocol {
+    private let credentialProvider: CredentialProviderProtocol
+    private let configuration: CacheClientConfigurationProtocol
+    private let controlClient: ControlClientProtocol
+    private let dataClient: DataClientProtocol
+    
+    init(
+        configuration: CacheClientConfigurationProtocol,
+        credentialProvider: CredentialProviderProtocol,
+        defaultTtlMillis: TimeInterval
```
Comment (malandis) [inline Sources/momento/CacheClient.swift:15]: Looks like `TimeInterval` is in seconds not in millis: https://developer.apple.com/documentation/foundation/timeinterval

**Review by malandis — COMMENTED** [review]
Looks good so far. I have a few minor comments around the tests where I focused my attention. I also found what looks like a typo regarding the default ttl.

If it's in scope, I would recommend adding documentation as you go now so you don't have to do it all at once later.

**Review by malandis — APPROVED** [review]
LGTM

File: Sources/momento/CacheClient.swift [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+    ```
+     switch response {
+     case let responseError as CacheSetError:
+        // handle error
+     case let responseSuccess as CacheSetSuccess:
+        // handle success
+     }
+    ```
+     */
+    public func set(cacheName: String, key: Data, value: Data, ttl: TimeInterval? = nil) async -> CacheSetResponse {
```
Comment (bruuuuuuuce) [inline Sources/momento/CacheClient.swift:285]: Just an observation I don't have any background in swift, but if its possible to have something like `key: Data | String` then we could reduce alot of the copy pasta'd code between all of these methods

File: Sources/momento/auth/CredentialProvider.swift [hunk 7]
```diff
@@ -6,15 +6,24 @@ enum CredentialProviderError: Error {
     case badToken
 }
 
+/// Specifies the fields that are required for a Momento client to connect to and authenticate with the Momento service.
 public protocol CredentialProviderProtocol {
+    /// API key provided by user, required to authenticate with the Momento service
     var originalAuthToken: String { get }
+    
+    /// parsed API key
```
Comment (bruuuuuuuce) [inline Sources/momento/auth/CredentialProvider.swift:14]: I am not grokking the diff between the `originalAuthToken` and the `authToken`

Comment (bruuuuuuuce) [reply Sources/momento/auth/CredentialProvider.swift:14]: Should these also not be `apiKeys`? Looks like we have moved most of the language over from authToken to apiKey

File: Sources/momento/internal/utils/Interceptors.swift [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+        [AuthHeaderInterceptor(apiKey: apiKey)]
+    }
+    
+    func makeListIndexesInterceptors() -> [GRPC.ClientInterceptor<ControlClient__ListIndexesRequest, ControlClient__ListIndexesResponse>] {
+        [AuthHeaderInterceptor(apiKey: apiKey)]
+    }
+}
+
+// Interceptor factory class for the CacheDataClient
+final class DataClientInterceptorFactory: CacheClient_ScsClientInterceptorFactoryProtocol {
```
Comment (bruuuuuuuce) [inline Sources/momento/internal/utils/Interceptors.swift:98]: this is very interesting, if I am reading this correctly, we are creating interceptors at the method level? So we have to define interceptors for every single method we define in the protocol

**Review by bruuuuuuuce — COMMENTED** [review]
Nice work looks good! Left a few comments/questions

File: Sources/momento/CacheClient.swift (hunk 6 above)
Comment (anitarua) [reply Sources/momento/CacheClient.swift:285]: That's a good point. Swift uses [enums for type unions](https://camlittle.com/posts/type-unions-in-swift/), I can try that approach out and reduce the amount of copy pasta

File: Sources/momento/internal/utils/Interceptors.swift (hunk 8 above)
Comment (anitarua) [reply Sources/momento/internal/utils/Interceptors.swift:98]: yes, and it wouldn't build if I excluded the currently un-implemented functions, so I had to include them all and decided I may as well have each of them return the correct auth header interceptor

Comment (bruuuuuuuce) [reply Sources/momento/internal/utils/Interceptors.swift:98]: Its an interesting approach swift took. On one hand, its super configurable and easy to define which methods get what interceptor, but its almost too configurable imo

Comment (bruuuuuuuce) [reply Sources/momento/internal/utils/Interceptors.swift:98]: 🤷

Comment (anitarua) [reply Sources/momento/internal/utils/Interceptors.swift:98]: 🤷🏻‍♀️

**Review by bruuuuuuuce — COMMENTED** [review]
Looks really good! I think this pattern is more scalable for our apis down the line. One nit I had was I'm not in love with this name for the data type `StringOrData`. I don't have a name I love, but here are some options I was thinking about

- StringData
- ScalarData
- ScalarType
- MomentoData

cc @malandis

**Comment by malandis** [comment]
@bruuuuuuuce @anitarua I favor names idiomatic to the language. In this case we should favor `data` over `bytes` since Swift has a native representation for bytes called `Data`. So the enum should be over `string` or `data`. Then the only thing left is the choice of name. `StringOrData` isn't very forward compatible with other types, so I favor a name like `ScalarType`.  A name like `ScalarData` is ambiguous with the Swift type `Data`.

**Review by malandis — APPROVED** [review]
Well done! 🎆

**Review by bruuuuuuuce — APPROVED** [review]
:shipit: awesome work!!!

**Pull request merged**